File: test_cases/net_003/create_testdata/net.py
import torch
import torch.nn as nn
import torch.nn.functional as F
import sys
import os
import pandas as pd

sys.path.insert(1, os.path.join(os.getcwd(), 'mkstd', "examples", "petab_sciml"))
sys.path.insert(1, os.path.join(os.getcwd(), 'test_cases'))
from petab_sciml_standard import Input, MLModel, PetabScimlStandard
from helper import read_array, get_ps_layer, get_dim, extract_numbers

class Net(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.layer1 = nn.Conv1d(1, 2, 5)
        self.layer2 = nn.Conv1d(2, 1, 5)

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        x = self.layer1(input)
        x = self.layer2(x)
        return x

# Create a pytorch module, convert it to PEtab SciML, then save it to disk.
dir_save = os.path.join(os.getcwd(), 'test_cases', "net_003")
net = Net()
mlmodel = MLModel.from_pytorch_module(
    module=net, mlmodel_id="model1", inputs=[Input(input_id="input1")]
)
petab_sciml_mlmodel = PetabScimlStandard.model(models=[mlmodel])
PetabScimlStandard.save_data(
    data=petab_sciml_mlmodel, filename=os.path.join(dir_save, "net.yaml")
)

# Test consistency between Lux.jl and PyTorch (a double check for correct values)
for i in range(1, 4):
    layer_names = ["layer1", "layer2", "layer3"]
    for layer_name in layer_names:
        df = pd.read_csv(os.path.join(dir_save, "net_ps_" + str(i) + ".tsv"), delimiter='\t')
        ps_weight = get_ps_layer(df, layer_name, "weight")
        ps_bias = get_ps_layer(df, layer_name, "bias")
        with torch.no_grad():
            layer = getattr(net, layer_name)
            layer.weight[:] = ps_weight
            layer.bias[:] = ps_bias

    df_input = pd.read_csv(os.path.join(dir_save, "net_input_" + str(i) + ".tsv"), delimiter='\t')
    df_output = pd.read_csv(os.path.join(dir_save, "net_output_" + str(i) + ".tsv"), delimiter='\t')
    input = read_array(df_input)
    output_ref = read_array(df_output)
    output = net.forward(input)
    torch.testing.assert_close(output_ref, output, atol=1e-3, rtol=0.0)

File: test_cases/net_029/create_testdata/net.jl
nn_model = @compact(
    layer1 = Dense(2, 5, Lux.gelu),
    layer2 = Dense(5, 1),
) do x
    embed = layer1(x)
    out = layer2(embed)
    @return out
end

for i in 1:3
    rng = StableRNG(i)
    ps, st = Lux.setup(rng, nn_model)
    input = rand(rng, 2)
    output = nn_model(input, ps, st)[1]
    df_ps = nn_ps_to_tidy(nn_model, ps, :net)
    df_input = _array_to_tidy(input)
    df_output = _array_to_tidy(output)
    CSV.write(joinpath(@__DIR__, "..", "net_ps_$i.tsv"), df_ps, delim = '\t')
    CSV.write(joinpath(@__DIR__, "..", "net_input_$i.tsv"), df_input, delim = '\t')
    CSV.write(joinpath(@__DIR__, "..", "net_output_$i.tsv"), df_output, delim = '\t')
end
write_yaml(joinpath(@__DIR__, ".."))
